Adds function to check by name if all required CRDs have been registered in Kubernetes

soenkeliebau opened the pull request:
The function also allows specifying a timeout and delay between retries to optionally allow operators to wait until CRDs have been created and then continue their startup.

partially fixes https://github.com/stackabletech/issues/issues/27

follow up PRs are needed in all operators for to fully fix the issue.

lfrancke commented on src/crd.rs:
```diff
@@ -60,6 +62,109 @@ where
+            Ok(()) => return Ok(()),
+            Err(err) => {
+                info!(
+                    "Error occurred when checking if all required CRDs are present: [{}]",
```
If I understand correctly this will show even for missing CRDs which does not feel like an "error" to me.

It'd be good if you could distinguish between actual errors and missing CRDs.
Errors should probably be `error!` as well while missing CRDs should probably print which CRDs are still missing (which this will do)

soenkeliebau replied:
True, I'll split that into two different match branches. 

Regarding the log level, I've followed the principle of logging at debug and info here and leaving anything more severe up to the code using this library. Otherwise we'll probably have duplicate logging once here and then the operator will log an error again.
That being said, this might be a bit of a special case since we probably won't return control to the operator here until all CRDs are present (unless the operator specifies a timeout and adds a loop of its own) so logging at warn might make sense..

lfrancke commented on tests/test_crd.rs:
```diff
@@ -65,3 +65,99 @@ async fn k8s_test_test_ensure_crd_created() {
+#[tokio::test]
+#[ignore = "Tests depending on Kubernetes are not ran by default"]
+async fn k8s_test_wait_for_crds() {
+    // TODO: Switch this to using TemporaryResource from the integration-test-commons crate
```
Every TODO should have a ticket (I'm not sure when the best time is for adding those references either)

lfrancke commented on tests/test_crd.rs:
```diff
@@ -65,3 +65,99 @@ async fn k8s_test_test_ensure_crd_created() {
+
+#[tokio::test]
+#[ignore = "Tests depending on Kubernetes are not ran by default"]
+async fn k8s_test_wait_for_crds() {
```
A test with two (or more) existing CRDs would be good to have

soenkeliebau commented:
I've changed the conflicting test cases to run serially and added a cleanup method that waits until resources are fully deleted by Kubernetes, which should hopefully resolve the flakiness.

lfrancke commented on src/client.rs:
```diff
@@ -323,6 +324,31 @@ impl Client {
+    ///
+    /// Afterwards it loops and checks regularly whether the resource has been deleted
+    /// from Kubernetes
+    pub async fn ensure_deleted<T>(&self, resource: T) -> OperatorResult<()>
```
Due to the nature of finalizers I believe we need a timeout here.

soenkeliebau replied:
I considered that, but we don't really offer timeouts in many other places and instead let users wrap the call in a timeout themselves if they need it.
I figured this was more in line with other fn calls - but am happy to add it too.
No real preference..

lfrancke replied:
Fair enough

lfrancke commented on src/crd.rs:
```diff
@@ -130,11 +130,34 @@ pub async fn wait_until_crds_present(
+                        return Err(RequiredCrdsMissing { names });
+                    }
+                }
+            }
```
Isn't this a duplicate now?
You handle this case below already, am I missing something?

soenkeliebau replied:
Nope, you are completely right. That was an earlier attempt at addressing your comment which I forgot to clean up.

lfrancke commented on src/client.rs:
```diff
@@ -345,7 +348,26 @@ impl Client {
+                    // if it does we'll log an error for now and continue iterating and wait for
+                    // the last interval we saw
+                    error!(
+                        "Waiting for deletion timed out, but no timeout was specified, this is an error and should not happen!"
```
Looking at ExponentialBackoff::default it defaults to "expire" after 15min: https://github.com/ihrwein/backoff/blob/master/src/default.rs#L13

I'm not sure if you want to change the docs or the behavior or whether having this backoff crate is overkill.
I'm fine with either of those solutions (or another one entirely)

soenkeliebau replied:
Ups, sorry, I missed that. I'll change the default behavior no never expire.

I considered whether we should add that crate, yes. But I pretty much reimplemented the exact functionality when I did it myself, just with less functionality. And I already found stuff that I wanted to add which would have duplicated even more of that crate.

I looked at the crate and it seemed more or less active, docs looked good and the author says where he took inspiration, which is a fairly established backoff algo by Google.

And with an eye on https://github.com/stackabletech/operator-rs/issues/147 I figured it might be worth pulling this in.
